### ISSUE TITLE
Does it support http proxy?

### ISSUE THREAD

**Issue by mao13820** [imapsync/imapsync#330]
I have set http proxy environment variables (https_proxy http_proxy). But I found it seems not work. When imapsync connect to mail server, such as gmail with port 993. It doesn't use the http proxy which I have set. I have checked my proxy service, it works well.

I wonder if imapsync support http proxy? If not, will you add this feature? Because you know, we can not directly access to some mail services (such as gmail) in some countries and regions.

Thank you very much.

**Comment by gilleslamiral** [comment]
Imapsync does imap and imaps, not http nor https.

**Comment by mao13820** [comment]
Yes, sure. But would it support socks5 proxy or tunneling tcp based http proxy? I tried many ways yesterday including proxychains,  graftcp. And I tried to modify the code about probe_imapssl (I faild when I try to add proxy to IO::Socket::SSL, it seems not support any proxy?). But all failed, do you have any idea about this? Thank you very much

**Comment by gilleslamiral** [comment]
>  do you have any idea about this?

What you can do is to use any imapsync proxy like the online service, which is an HTTPS service. 
For example, the basic imapsync example:

````
  imapsync \
     --host1 test1.lamiral.info --user1 test1 --password1 secret1 \
     --host2 test2.lamiral.info --user2 test2 --password2 secret2
````

can be done with curl:

     curl -v --data 'host1=test1.lamiral.info;user1=test1;password1=secret1;host2=test2.lamiral.info;user2=test2;password2=secret2' https://i014.lamiral.info/cgi-bin/imapsync

curl supports http_proxy but you may not need it since https is usually allowed from anywhere.

````
man curl
...
ENVIRONMENT
       The environment variables can be specified in lower case or upper case. 
       The lower case version has precedence. http_proxy is an exception as it is only
       available in lower case.

       Using an environment variable to set the proxy has the same effect as using the --proxy option.

       http_proxy [protocol://]<host>[:port]
              Sets the proxy server to use for HTTP.

       HTTPS_PROXY [protocol://]<host>[:port]
              Sets the proxy server to use for HTTPS.
````

**Comment by mao13820** [comment]
Yes, curl support http proxy. But this method seems rely on third-party website. I wonder if there is any method to let the imapsync itself support proxy? Socks 5 or HTTP. Thank you very much.

**Comment by gilleslamiral** [comment]
> But this method seems to rely on a third-party website

It can be your website, your proxy, if you install one:
https://imapsync.lamiral.info/INSTALL.d/INSTALL.OnlineUI.txt

>  I wonder if there is any method to let the imapsync itself support proxy? Socks 5 or HTTP

We are back to my first response. 
Imapsync does imap and imaps because imap servers talk imap or imaps.
Let's take the Gmail imap server as an example.
Gmail is accessible via the protocol IMAPS. 
How technically can we access it with the protocol HTTP or Sock5 ?

- HTTP: with an imapsync proxy like the online https://imapsync.lamiral.info/cgi-bin/imapsync
- Sock5: with a Socksifier like tun2socks https://github.com/xjasonlyu/tun2socks/wiki/Examples

**Comment by mao13820** [comment]
Thank you for your time and patience.

Yes. So I wonder if there is a way that I can implement a proxy (TCP proxy through socks5 or http connect proxy) inside the program.

Because if not, we can only make the host in a proxy environment to achieve the goal.
1. Make the host out of the limited network (use imapsync online service). 
2. Implement a vpn on the host (tun2socks).

The problem is that these two methods is too heavy and unconvenient. I wonder if there is a way to implement proxy function inside the program itself, so we can only use the program to solve the network problem other than got a new server or modify the host network (creating a tun/tap and modifying the router table). The better thing is that the program can be implemented with docker.

So do you have any idea about how to realise proxy inside this program? I want to achieve this but I don't know how to start.

Thank you again.

**Comment by gilleslamiral** [comment]
Can you explain a little more your context?

An imap server like Gmail talks IMAP, it does not talk HTTP, at all. 
An HTTP proxy talks HTTP, not imap.

How do think an HTTP proxy can talk to Gmail?

**Comment by gilleslamiral** [comment]
Can you explain a little more your context?

An imap server like Gmail talks IMAP, it does not talk HTTP, at all. 
An HTTP proxy talks HTTP, not imap.

How do think an HTTP proxy can talk to Gmail?

**Comment by mao13820** [comment]
Thank you for your attetion to my issue.

According to 

RFC 7231, section 4.3.6: CONNECT
https://datatracker.ietf.org/doc/rfc7231/

We can use HTTP CONNECT provided by HTTP PROXY SERVER to proxy any TCP connection. IMAP is based on TCP. So it should work.

There are many cases of implement, such as proxychains, proxytunnel, graftcp

https://github.com/haad/proxychains
https://github.com/proxytunnel/proxytunnel
https://github.com/hmgle/graftcp

It all can be used for proxy any TCP connection through HTTP PROXY, including SSH, IMAP, SMTP, TLS, etc.

I have tried proxytunnel to solve this problem and successed today. You can see its documents following here which explain how it works.

https://github.com/proxytunnel/proxytunnel/blob/1026053ed2acfb2b990f96ee9f1b76342e1b4bb1/docs/proxytunnel-paper.adoc#using-socks

And from the document, I found why I failed to proxy imapsync by using these kind of tools (like proxychains above) through HTTP PROXY before.  Because my http proxy server was not permitted to access any mail service port (like 993) on other server. And I fixed it, then I tried proxychains and proxytunnel and all success.

So I wonder how to implement the http or socks5 proxy inside imapsync? Looking forward to your advice. 

Thank you again.

**Comment by mao13820** [comment]
I have found a way to solve this problem.
If add these before any use

use IO::Socket::Socks::Wrapper (
{
    ProxyAddr => 'x.x.x.x',
    ProxyPort => xxx,
}
);

The imapsync will use the this socks5 to proxy connection

But as for http proxy, I am still trying to find a solution

**Comment by gilleslamiral** [comment]
> I have found a way to solve this problem.
> use IO::Socket::Socks::Wrapper (
> ...

Have a go with this solution!
Try it on imapsync, it's also Perl, you have the source code and the hand on a Socks server,
I don't have both.

**Comment by gilleslamiral** [comment]
> I have tried proxytunnel to solve this problem and succeeded today.

That's all enough to my ears. 
I won't add proxytunnel inside imapsync when they already work well together.

**Comment by mao13820** [comment]
> > I have found a way to solve this problem.
> > use IO::Socket::Socks::Wrapper (
> > ...
> 
> Have a go with this solution! Try it on imapsync, it's also Perl, you have the source code and the hand on a Socks server, I don't have both.

Yes, of course. I have tried it at that time and it success as I said. But this method only work with socket5 proxy, not http proxy. Http proxy solution could be using tools like proxychains and proxytunnel, etc which I mentioned above.

I totally understand your thought. You can for sure decide which function you want to add in imapsync. I just want to share the problem I encountered and how I solve it.

Please believe that I didn't mean to offend you and I really thank you for offering so great tools for us.

**Comment by gilleslamiral** [comment]
> I totally understand your thought. You can for sure decide which function you want to add in imapsync. I just want to share the problem I encountered and how I solve it.

Thanks for your input! 

If you want to make a better document than this discussion, feel free to write it, I will publish it in the FAQ.d directory:
https://imapsync.lamiral.info/FAQ.d/
and summarise it at
https://imapsync.lamiral.info/#doc

**Comment by mao13820** [comment]
Yes, I do would like to contribute to this project. I'll do some research deeper in my spare time.
